src/Kestrel.Core/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -181,6 +180,22 @@ public Task WriteStreamSuffixAsync()
             }
         }
 
+        public Task WriteRstStreamAsync(bool endReceived, Http2ErrorCode error)
+        {
+            lock (_dataWriterLock)
+            {
+                // Only reset if the connection is open or in a half closed state.
+                if (endReceived && _completed)
+                {
+                    return Task.CompletedTask;
+                }
+
+                _completed = true;
```

Comment:
If you set `_completed` to true, you're responsible for completing the data PipeWriter. I would probably just call Dispose() here.